src/Builder/Encoder/OperatorEncoder.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MongoDB\Builder\Encoder;
+
+use LogicException;
+use MongoDB\Builder\Stage\GroupStage;
+use MongoDB\Builder\Type\Encode;
+use MongoDB\Builder\Type\OperatorInterface;
+use MongoDB\Builder\Type\Optional;
+use MongoDB\Codec\EncodeIfSupported;
+use MongoDB\Exception\UnsupportedValueException;
+use stdClass;
+
+use function array_key_exists;
+use function assert;
+use function get_object_vars;
+use function is_array;
+use function is_object;
+use function property_exists;
+use function sprintf;
+
+/** @template-extends AbstractExpressionEncoder<stdClass, OperatorInterface> */
+class OperatorEncoder extends AbstractExpressionEncoder
+{
+    /** @template-use EncodeIfSupported<stdClass, OperatorInterface> */
+    use EncodeIfSupported;
+
+    public function canEncode(mixed $value): bool
+    {
+        return $value instanceof OperatorInterface;
+    }
+
+    public function encode(mixed $value): stdClass
+    {
+        if (! $this->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
+        }
+
+        switch ($value::ENCODE) {
+            case Encode::Single:
+                return $this->encodeAsSingle($value);
+
+            case Encode::Array:
+                return $this->encodeAsArray($value);
+
+            case Encode::Object:
+                return $this->encodeAsObject($value);
+
+            case Encode::DollarObject:
+                return $this->encodeAsDollarObject($value);
+
+            case Encode::Group:
+                assert($value instanceof GroupStage);
+
+                return $this->encodeAsGroup($value);
+        }
+
+        throw new LogicException(sprintf('Class "%s" does not have a valid ENCODE constant.', $value::class));
+    }
+
+    /**
+     * Encode the value as an array of properties, in the order they are defined in the class.
+     */
+    private function encodeAsArray(OperatorInterface $value): stdClass
+    {
+        $result = [];
+        /** @var mixed $val */
+        foreach (get_object_vars($value) as $val) {
+            // Skip optional arguments.
+            // $slice operator has the optional <position> argument in the middle of the array
```

Comment:
```suggestion
            // Skip optional arguments. For example, the $slice expression operator has an optional <position> argument
            // in the middle of the array.
```

Rewriting this for clarity. I forgot that it was the _expression_ operator. Alternatively, we could refer to the actual `MongoDB\Builder\Expression\SliceOperator` class since that's where `Optional::Undefined` is used, but it'd be less concise.